src/mods/storytimer.cpp

```diff
@@ -0,0 +1,368 @@
+         mkb::sub_mode == mkb::SMD_GAME_RINGOUT_MAIN || mkb::sub_mode == mkb::SMD_GAME_RETRY_INIT ||
+         mkb::sub_mode == mkb::SMD_GAME_RETRY_MAIN);
+
+    // submodes during timeover
+    bool is_timeover = (mkb::sub_mode == mkb::SMD_GAME_TIMEOVER_INIT ||
+                        mkb::sub_mode == mkb::SMD_GAME_TIMEOVER_MAIN);
+
+    // this code is used to halt the timer once the screen becomes completely white when stage
+    // selecting out of a level
+    if (mkb::pausemenu_type == mkb::PMT_STORY_PLAY &&
+        mkb::g_current_focused_pause_menu_entry == 4 && pad::button_pressed(mkb::PAD_BUTTON_A) &&
+        is_postgoal) {
+        // stage select is on line 4 of the pause menu (top line is line 0)
+        s_start_stage_fade_out_timer = true;
+    } else if (mkb::g_storymode_stageselect_state == mkb::STAGE_SELECT_INTRO_SEQUENCE) {
+        s_start_stage_fade_out_timer = false;
+    }
+    if (s_start_stage_fade_out_timer) {
+        s_stage_fade_out_timer += 1;
+    } else {
+        s_stage_fade_out_timer = 0;
+    }
+
+    for (s32 k = 0; k < 10; k++) {
+        if (mkb::scen_info.world == k) {
+            if (mkb::sub_mode == mkb::SMD_GAME_FIRST_INIT) {
+                // need to add 1 additional frame to the timer during spin in
+                // putting this code before the code below for s_timer_group[k].full_world makes
+                // the timer tick up more naturally when transitioning from the 10 ball screen
+                // to spin in, more specifically, this prevents the timer from skipping ahead
+                // for a few frames, then pausing for a few frames to even itself out (I don't
+                // understand why that happens)
+                s_timer_group[k].non_gameplay += 1;
+            }
+
+            s_timer_group[k].full_world = s_timer_group[k].gameplay +
+                                          s_timer_group[k].non_gameplay +
+                                          s_timer_group[k].world_start_correction;
+
+            if (s_completed_stages_world[k] <= 9) {
+                s_timer_group[k].segment = s_timer_group[k].full_world;
+            } else if (s_completed_stages_world[k] == 10 &&
+                       mkb::sub_mode == mkb::SMD_GAME_GOAL_INIT && !paused_now) {
+                // stop world k's segment timer on tape break of the last stage of the world; we
+                // need to correct by 2f since SMD_GAME_GOAL_INIT happens 2f after tape break
+                s_timer_group[k].segment += -2;
+            }
+
+            if (is_on_spin_in || is_on_exit_game || is_on_fallout || is_timeover) {
+                // increment the timer during every frame on spin in, exit game, fallout, and
+                // timeover
+                s_timer_group[k].non_gameplay += 1;
+            }
+
+            if (is_on_gameplay) {
+                // increment the timer every frame during gameplay
+                s_timer_group[k].gameplay += 1;
```

Comment:
Technically the tape-rip frame and the frame after will both be counted for gameplay time, would you want to correct that since it'd be a bonus 2 frames counted per stage?